[3.2] Fix multiple issues with one-way collisions

madmiraal opened the pull request:
3.2 version of #42574, i.e a salvage of #40645.

madmiraal committed:
Fix multiple issues with one-way collisions

For RigidBodies, uses the collision normal determined by relative motion
to determine whether or not a one-way collision has occurred.

For KinematicBodies, performs additional checks to ensure a one-way
collision has occurred, and averages the recovery step over all collision
shapes.

Co-authored-by:    Sergej Gureev <sergej.gureev@relex.fi>

capnm commented on servers/physics_2d/physics_2d_server_sw.cpp:
```diff
@@ -167,23 +167,14 @@ void Physics2DServerSW::_shape_col_cbk(const Vector2 &p_point_A, const Vector2 &
 		if (p_point_A.distance_squared_to(p_point_B) > cbk->valid_depth * cbk->valid_depth) {
 			cbk->invalid_by_dir++;
 			return;
 		}
```
?
```c++
	Vector2 rel_dir = p_point_A - p_point_B;
	if (rel_dir.length_squared() > 0) {
		if (cbk->valid_depth < 10e20 && cbk->valid_dir != Vector2()) {
			if (p_point_A.distance_squared_to(p_point_B) > cbk->valid_depth * cbk->valid_depth) {
				cbk->invalid_by_dir++;
				return;
			}
			if (cbk->valid_dir.dot(rel_dir.normalized()) <= 0) {
				cbk->invalid_by_dir++;
				return;
			}
		}
	}
```

madmiraal replied:
I agree that optimization is possible. My goal in this PR is to keep the changes to a minimum, which makes it easier to review, and, therefore, more likely to be merged.

capnm replied:
It's the same, personally I just prefer to avoid normalizing null vectors as much as possible as it often leads to errors that are difficult to spot.

madmiraal replied:
Updated.

RandomShaper commented:
I'm approving this for the same reasons explained here: https://github.com/godotengine/godot/pull/42574#issuecomment-711046468. (In this case, considering 3.2 is the production release, the part about testing a project which uses one-way heavily is much more important.)

madmiraal commented:
Updated with the changes applied to #42574 i.e. both KinematicBody and RigidBody treat perpendicular contacts as pass through; and removed the change the was causing the bug.

madmiraal commented:
Updated to include a fix for the bugs identified by @Rhathe [here](https://github.com/godotengine/godot/pull/42574#issuecomment-715392217) and [here](https://github.com/godotengine/godot/pull/42574#issuecomment-716117102).

madmiraal commented:
Updated to properly fix the bugs identified by @Rhathe and #43012.

madmiraal commented:
Updated to include the fix for #43266, identified by @AttackButton [here](https://github.com/godotengine/godot/pull/42574#issuecomment-719957095).

madmiraal commented:
Updated to ensure the work around applied in the previous update only applies during the unsticking phase.

madmiraal commented:
Updated to include the small refactor that removed some unneeded code.

madmiraal commented:
Updated to include the changes applied to #42574:
1. Tolerances for dot product direction calculations.
2. Applied the division by `cbk.amount` for recovery motion to `test_body_ray_separation` and to both in Godot 3D too for consistency.

akien-mga commented:
Thanks!
